core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -1188,28 +1221,32 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
 
               cur = null
 
-              while ((cur eq null) && check()) {
+              while (cont && (cur eq null) && check()) {
                 runLoop(iterate.body(), stackIndex, stackIndex, nextDepth) match {
                   case Success(value) =>
                     iterate.process(value)
 
                   case failure =>
+                    cont = failure ne null
```

Comment:
Let's be explicit here and use `case null => cont = false` instead

Reply:
this might have performance implications, can we settle for a verbose comment?